src-extras/Database/LSMTree/Extras/MergingRunData.hs

```diff
@@ -0,0 +1,217 @@
+  deriving stock (Show, Eq)
+
+-- | See @mergeInvariant@ in the prototype.
+mergingRunDataInvariant :: MergingRunData t k v b -> Either String ()
+mergingRunDataInvariant = \case
+    CompletedMergeData _ (MR.NumRuns n) _ ->
+      assertI "completed merges are non-trivial (at least two inputs)" $
+        n > 1
+    OngoingMergeData _ rds -> do
+      assertI "inputs to ongoing merges aren't empty" $
+        all nonEmptyRunDataInvariant rds
+      assertI "ongoing merges are non-trivial (at least two inputs)" $
+        length rds > 1
+  where
+    assertI msg False = Left msg
+    assertI _   True  = Right ()
+
+mapMergingRunData ::
+     Ord k'
+  => (k -> k') -> (v -> v') -> (b -> b')
+  -> MergingRunData t k v b -> MergingRunData t k' v' b'
+mapMergingRunData f g h = \case
+    CompletedMergeData t n r ->
+      CompletedMergeData t n $ mapRunData f g h r
+    OngoingMergeData t rs ->
+      OngoingMergeData t $
+        map (NonEmptyRunData . mapRunData f g h . unNonEmptyRunData) rs
+
+type SerialisedMergingRunData t =
+    MergingRunData t SerialisedKey SerialisedValue SerialisedBlob
+
+serialiseMergingRunData ::
+     (SerialiseKey k, SerialiseValue v, SerialiseValue b)
+  => MergingRunData t k v b -> SerialisedMergingRunData t
+serialiseMergingRunData =
+    mapMergingRunData serialiseKey serialiseValue serialiseBlob
+
+{-------------------------------------------------------------------------------
+  QuickCheck
+-------------------------------------------------------------------------------}
+
+labelMergingRunData ::
+     Show t => SerialisedMergingRunData t -> Property -> Property
+labelMergingRunData (CompletedMergeData mt _ rd) =
+      tabulate "merging run state" ["CompletedMerge"]
+    . tabulate "merge type" [show mt]
+    . labelRunData rd
+labelMergingRunData (OngoingMergeData mt rds) =
+      tabulate "merging run state" ["OngoingMerge"]
+    . tabulate "merge type" [show mt]
+    . tabulate "merging run inputs" [showPowersOf 2 (length rds)]
+    . foldr ((.) . labelRunData . unNonEmptyRunData) id rds
+
+instance ( Arbitrary t, Ord k, Arbitrary k, Arbitrary v, Arbitrary b
+         ) => Arbitrary (MergingRunData t k v b) where
+  arbitrary = genMergingRunData arbitrary arbitrary arbitrary
+  shrink = shrinkMergingRunData shrink shrink shrink
+
+genMergingRunData ::
+     forall t k v b. (Arbitrary t, Ord k)
+  => Gen k
+  -> Gen v
+  -> Gen b
+  -> Gen (MergingRunData t k v b)
+genMergingRunData genKey genVal genBlob =
+    QC.oneof
+      [ do
+          mt <- arbitrary
+          numRuns <- MR.NumRuns <$> QC.chooseInt (2, 32)
+          rd <- genRunData genKey genVal genBlob
+          pure (CompletedMergeData mt numRuns rd)
+      , do
+          mt <- arbitrary
+          n  <- QC.chooseInt (2, 8)
+          rs <- QC.vectorOf n $ QC.scale (`div` n) $
+            -- scaled, so overall size is similar to a single RunData
+            genNonEmptyRunData genKey genVal genBlob
```

Comment:
We should check, but I think this not only scales the number of key/value pairs, but also the size of the keys/values themselves. Is that problematic?

Reply:
*has large k/op* is not the only affected metric, in general everything just gets smaller, even keys

Something like this would work:
```haskell
          rs <- QC.vectorOf n $ do
            s <- getSize
            QC.scale (`div` n) $
            -- scaled, so overall size is similar to a single RunData
            genNonEmptyRunData (resize s genKey) (resize s genVal) (resize s genBlob)
```
Not saying this is definitely necessarily, but it's relatively simple, no?

Reply:
Ah, true, that's pretty straightforward.